components/node-labeler/cmd/run.go

```diff
@@ -249,6 +282,184 @@ func (r *PodReconciler) Reconcile(ctx context.Context, req reconcile.Request) (r
 	return reconcile.Result{}, nil
 }
 
+type WorkspaceCountController struct {
+	client.Client
+	nodesToReconcile chan string
+	stopChan         chan struct{}
+}
+
+func NewWorkspaceCountController(client client.Client) (*WorkspaceCountController, error) {
+	return &WorkspaceCountController{
+		Client:           client,
+		nodesToReconcile: make(chan string, 1000),
+		stopChan:         make(chan struct{}),
+	}, nil
+}
+
+func (wc *WorkspaceCountController) SetupWithManager(mgr ctrl.Manager) error {
+	// Start the periodic reconciliation goroutine
+	go wc.periodicReconciliation()
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("workspace-count").
+		For(&workspacev1.Workspace{}).
+		WithEventFilter(wc.workspaceFilter()).
+		Complete(wc)
+}
+
+func (wc *WorkspaceCountController) periodicReconciliation() {
+	ticker := time.NewTicker(5 * time.Minute)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			log.Info("Starting periodic full reconciliation")
+			ctx := context.Background()
+			if _, err := wc.reconcileAllNodes(ctx); err != nil {
+				log.WithError(err).Error("periodic reconciliation failed")
+			}
+		case <-wc.stopChan:
+			return
+		}
+	}
+}
+
+func (wc *WorkspaceCountController) workspaceFilter() predicate.Predicate {
+	return predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			ws := e.Object.(*workspacev1.Workspace)
+			if ws.Status.Runtime == nil {
+				log.WithField("workspace", ws.Name).Info("workspace not ready yet")
+				return false
+			}
+
+			return ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != ""
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			wsOld := e.ObjectOld.(*workspacev1.Workspace)
+			ws := e.ObjectNew.(*workspacev1.Workspace)
+			// if we haven't seen runtime info before and now it's there, let's reconcile
+			if wsOld.Status.Runtime == nil && ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != "" {
+				return true
+			}
+
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			ws := e.Object.(*workspacev1.Workspace)
+			if ws.Status.Runtime != nil && ws.Status.Runtime.NodeName != "" {
+				// Queue the node for reconciliation
+				select {
+				case wc.nodesToReconcile <- ws.Status.Runtime.NodeName:
```

Comment:
When deleting, we won't be able to query the workspaces' node name, since it already won't exist at the time we'll query for it. Because of that, we rather capture node names in a channel, which is consumed any time `Reconcile` is run.